### PR TITLE
[feature](datetime) "timediff" supports calculating microseconds

### PR DESCRIPTION
**Pull request by Mryange**
## Proposed changes
In the previous code, we distinguished between two types of time, "time" and "timev2", but on the backend, both of them are the same.
So here, we are making some modifications. The unit for "time" is seconds, while the unit for "timev2" is microseconds.
Issue Number:  https://github.com/apache/doris/issues/17174
Due to the mixed usage of "time" and "timev2" in many places in the past, this pull request will be **converted to a draft** first for regression testing.

before
```
mysql [test]>select timediff(
cast('0001-01-01 00:00:00.114514' as Datetimev2(6) ) , 
cast('0001-01-01 00:00:00.000000' as Datetimev2(6) ));
+--------------------------------------------------------------------------------------------------------------------+
| timediff(CAST('0001-01-01 00:00:00.114514' AS DATETIMEV2(6)), CAST('0001-01-01 00:00:00.000000' AS DATETIMEV2(6))) |
+--------------------------------------------------------------------------------------------------------------------+
| 00:00:00                                                                                                   |
+--------------------------------------------------------------------------------------------------------------------+
```
after
```
mysql [test]>select timediff(
cast('0001-01-01 00:00:00.114514' as Datetimev2(6) ) , 
cast('0001-01-01 00:00:00.000000' as Datetimev2(6) ));
+--------------------------------------------------------------------------------------------------------------------+
| timediff(CAST('0001-01-01 00:00:00.114514' AS DATETIMEV2(6)), CAST('0001-01-01 00:00:00.000000' AS DATETIMEV2(6))) |
+--------------------------------------------------------------------------------------------------------------------+
| 00:00:00.114514                                                                                                    |
+--------------------------------------------------------------------------------------------------------------------+
mysql [(none)]>select timediff("2010-01-01 01:00:00.123", "2010-01-05 01:00:00.4567");
+-----------------------------------------------------------------+
| timediff('2010-01-01 01:00:00.123', '2010-01-05 01:00:00.4567') |
+-----------------------------------------------------------------+
| -96:00:00.3337                                                  |
+-----------------------------------------------------------------+
```

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 41.38 seconds
 stream load tsv:          457 seconds loaded 74807831229 Bytes, about 156 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          56 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          28 seconds loaded 861443392 Bytes, about 29 MB/s
 insert into select:          67.9 seconds inserted 10000000 Rows, about 147K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230630090149_clickbench_pr_170376.html

**Comment by Mryange** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 38.3 seconds
 stream load tsv:          456 seconds loaded 74807831229 Bytes, about 156 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          57 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          28 seconds loaded 861443392 Bytes, about 29 MB/s
 insert into select:          69.9 seconds inserted 10000000 Rows, about 143K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230702032227_clickbench_pr_170924.html

File: fe/fe-core/src/main/java/org/apache/doris/nereids/types/TimeV2Type.java [hunk 1]
```diff
@@ -28,13 +29,26 @@ public class TimeV2Type extends PrimitiveType {
...
 
     private static final int WIDTH = 8;
+    private final int scale;
+
+    private TimeV2Type(int scale) {
+        this.scale = scale;
+    }
 
     private TimeV2Type() {
+        scale = 6;
```
Comment (Gabriel39) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/types/TimeV2Type.java:39]: Why is 6?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/TimeDiff.java [hunk 2]
```diff
@@ -46,8 +48,7 @@ public class TimeDiff extends ScalarFunction
             FunctionSignature.ret(TimeType.INSTANCE).args(DateTimeType.INSTANCE, DateTimeType.INSTANCE),
             FunctionSignature.ret(TimeV2Type.INSTANCE).args(DateTimeV2Type.SYSTEM_DEFAULT, DateV2Type.INSTANCE),
             FunctionSignature.ret(TimeV2Type.INSTANCE).args(DateV2Type.INSTANCE, DateTimeV2Type.SYSTEM_DEFAULT),
-            FunctionSignature.ret(TimeV2Type.INSTANCE).args(DateV2Type.INSTANCE, DateV2Type.INSTANCE)
-    );
+            FunctionSignature.ret(TimeType.INSTANCE).args(DateV2Type.INSTANCE, DateV2Type.INSTANCE));
```
Comment (Gabriel39) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/TimeDiff.java:51]:
```suggestion
            FunctionSignature.ret(TimeV2Type.INSTANCE).args(DateV2Type.INSTANCE, DateV2Type.INSTANCE));
```

File: fe/fe-core/src/main/java/org/apache/doris/analysis/FunctionCallExpr.java [hunk 3]
```diff
@@ -1559,6 +1559,13 @@ && collectChildReturnTypes()[0].isDecimalV3()) {
         }
         if (fn.getFunctionName().getFunction().equals("timediff")) {
             fn.getReturnType().getPrimitiveType().setTimeType();
+            ScalarType left = (ScalarType) argTypes[0];
+            ScalarType rigth = (ScalarType) argTypes[1];
+            if (left.isDatetimeV2() || rigth.isDatetimeV2()) {
```
Comment (Gabriel39) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/FunctionCallExpr.java:1564]: also for datev2

File: fe/fe-common/src/main/java/org/apache/doris/catalog/ScalarType.java [hunk 4]
```diff
@@ -432,7 +434,7 @@ public static ScalarType createTimeType() {
         if (!Config.enable_date_conversion) {
             return new ScalarType(PrimitiveType.TIME);
         }
-        ScalarType type = new ScalarType(PrimitiveType.TIMEV2);
+        ScalarType type = new ScalarType(PrimitiveType.TIME);
```
Comment (Gabriel39) [inline fe/fe-common/src/main/java/org/apache/doris/catalog/ScalarType.java:437]: Why make this change?

**Comment by Mryange** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 39.55 seconds
 stream load tsv:          455 seconds loaded 74807831229 Bytes, about 156 MB/s
 stream load json:         21 seconds loaded 2358488459 Bytes, about 107 MB/s
 stream load orc:          58 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          29 seconds loaded 861443392 Bytes, about 28 MB/s
 insert into select:          68.9 seconds inserted 10000000 Rows, about 145K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230704060810_clickbench_pr_171826.html

**Comment by Mryange** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 41.74 seconds
 stream load tsv:          465 seconds loaded 74807831229 Bytes, about 153 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          57 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          29 seconds loaded 861443392 Bytes, about 28 MB/s
 insert into select:          70.1 seconds inserted 10000000 Rows, about 142K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230704071924_clickbench_pr_171869.html

**Comment by Mryange** [comment]
run p0

**Review by Gabriel39 — DISMISSED** [review]
LGTM

**Comment by Mryange** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 48.19 seconds
 stream load tsv:          454 seconds loaded 74807831229 Bytes, about 157 MB/s
 stream load json:         21 seconds loaded 2358488459 Bytes, about 107 MB/s
 stream load orc:          57 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          28 seconds loaded 861443392 Bytes, about 29 MB/s
 insert into select:          70.4 seconds inserted 10000000 Rows, about 142K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230709064349_clickbench_pr_175087.html

**Comment by Mryange** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 47.01 seconds
 stream load tsv:          441 seconds loaded 74807831229 Bytes, about 161 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          56 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          28 seconds loaded 861443392 Bytes, about 29 MB/s
 insert into select:          25.8 seconds inserted 10000000 Rows, about 387K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230709144759_clickbench_pr_175273.html

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 55.11 seconds
 stream load tsv:          501 seconds loaded 74807831229 Bytes, about 142 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          28.5 seconds inserted 10000000 Rows, about 350K ops/s
 storage size: 17170782441 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230709225130_clickbench_pr_175274.html

**Comment by Mryange** [comment]
run arm

**Review by Gabriel39 — DISMISSED** [review]
LGTM

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 58.52 seconds
 stream load tsv:          501 seconds loaded 74807831229 Bytes, about 142 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.1 seconds inserted 10000000 Rows, about 343K ops/s
 storage size: 17162577350 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230710133829_clickbench_pr_175710.html

**Comment by Mryange** [comment]
run p0

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 47.68 seconds
 stream load tsv:          447 seconds loaded 74807831229 Bytes, about 159 MB/s
 stream load json:         21 seconds loaded 2358488459 Bytes, about 107 MB/s
 stream load orc:          56 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          29 seconds loaded 861443392 Bytes, about 28 MB/s
 insert into select:          26.1 seconds inserted 10000000 Rows, about 383K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230710080229_clickbench_pr_175707.html

**Review by starocean999 — APPROVED** [review]

**Review by Gabriel39 — APPROVED** [review]
LGTM

**Pull request merged**